File: src/main/java/com/github/urlshorten/UrlShortenController.java
package com.github.urlshorten;

import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.net.URI;

@RestController
@Validated
public class UrlShortenController {

    private final UrlShortenService urlShortenService;
    private final Mapper mapper;

    public UrlShortenController(UrlShortenService urlShortenService, Mapper mapper) {
        this.urlShortenService = urlShortenService;
        this.mapper = mapper;
    }

    @PostMapping("/url_shorten")
    public ResponseEntity shorten(@RequestBody UrlToShorten url) {
        var shortenedUrl = urlShortenService.shorten(url.getUrl(), mapper.toIdStrategy(url.getStrategy()));

        return ResponseEntity.created(URI.create(shortenedUrl.getShortenUrl())).build();
    }

    @GetMapping("/url_shorten/{url}")
    public ResponseEntity<ShortenedUrl> retrieve(@PathVariable("url") String url) {
        return ResponseEntity.of(urlShortenService.retrieve(url).map(mapper::toDto));
    }
}

File: src/main/kotlin/com/github/urlshorten/UrlShortenController.kt
package com.github.urlshorten

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import java.net.URI

@RestController
@Validated
class UrlShortenController(val service: UrlShortenService) {

    @GetMapping("/url_shorten/{url}")
    fun retrieve(@PathVariable("url") url: String): ResponseEntity<ShortenedUrl> {
        val retrievedUrl = service.retrieve(url)
        return if (retrievedUrl == null) {
            ResponseEntity<ShortenedUrl>(null, HttpStatus.NOT_FOUND)
        } else ResponseEntity.ok(retrievedUrl.let {
            ShortenedUrl(
                    shortenUrl = it.shortenUrl ?: "",
                    originalUrl = it.originalUrl,
                    creationDate = it.creationDate,
                    expirationDate = it.expirationDate,
            )
        })
    }

    @PostMapping("/url_shorten")
    fun shorten(@RequestBody url: UrlToShorten) = ResponseEntity.created(URI.create(service.shorten(url.url, IdStrategy.valueOf(url.strategy.name))?.shortenUrl)).build<URI>()
}